### PR TITLE
Separate Threads from Position

### PR DESCRIPTION
**Pull request by lucasart**
Threads don't belong in Position. This simplifies the Position API, and makes more easily separable from the rest of Stockfish (for those using the Position API it in other programs or experiments).

The "trick" is to make ThisThread a thread_local variable. That makes it a lot easier to access from anywhere in the code (you don't need a Position object).

No functional change.

On my configuration (i7-3770k, GCC 5.2.1, Linux 4.2.0), this is actually a small speed-up:
```
stat       test     master    diff
mean  2,600,002  2,589,240  12,490
stdev    15,077     14,466   6,771

speedup       0.48%
P(speedup>0) 100.0%
```

### CODE REVIEW

**Comment by joergoster** [comment]
Looks like a small loss of speed on my box.
```
Engine                          | Nodes/second
sf-thisThread                   | 1644456.0 +- 3468.12
sf-master                       | 1684573.0 +- 6867.68

Differences                     | -40465.0 +- 1140.0
Variance of the mean            | 161.22 ( -0.4 %)
Speed up                        | -2.4 %
```
Configuration: AMD FX-8120@4GHz, gcc 5.3.0, Linux 3.16), make build ARCH=x86-64-modern, strip stockfish

However, this should be tested for no regression, shouldn't it?

**Comment by lucasart** [comment]
You  should  **always** do a profile build when  doing  speed  testing (or any  serious  testing).

**Comment by joergoster** [comment]
I *knew* you would be coming up with this.
Comparison of 2 profile-builds of the master:
```
Engine                          | Nodes/second
sf-master-pb2                   | 1928782.0 +- 3700.58
sf-master-pb1                   | 1938852.0 +- 2542.92

Differences                     | -9850.0 +- 1572.0
Variance of the mean            | 222.31 ( -2.26 %)
Speed up                        | -0.51 %
```
Which one to choose?

Comparison of 2 profile-builds of your test branch:
```
Engine                          | Nodes/second
sf-this-pb2                     | 1907426.5 +- 4246.48
sf-this-pb1                     | 1908080.0 +- 3330.36

Differences                     | 0.0 +- 2002.0
Variance of the mean            | inf ( inf %)
Speed up                        | 0.0 %
```
Accidently equal in terms of speed. Please note, this is quite unusual. In general, you will get differences with each profile build.


Comparison 1:
```
Engine                          | Nodes/second
sf-this-pb1                     | 1908516.0 +- 3332.84
sf-master-pb1                   | 1938627.0 +- 4304.08

Differences                     | -30183.0 +- 1401.0
Variance of the mean            | 198.13 ( -0.66 %)
Speed up                        | -1.56 %
```

Comparison 2:
```
Engine                          | Nodes/second
sf-this-pb1                     | 1908952.0 +- 2030.18
sf-master-pb2                   | 1928782.0 +- 4399.8

Differences                     | -20057.0 +- 2339.0
Variance of the mean            | 330.78 ( -1.65 %)
Speed up                        | -1.04 %
```

Of course, you must **NEVER** use profile builds. Any serious testing should be done with normal builds. At least, we don't want to measure possible enhancements/speedups of the compiler. Do we?
In my opinion, this should only be the last step for a release version.

@glinscott @mcostalba @zamar What's the opinion of the other maintainers? Do you still think it was a good decision to use profile builds in fishtest?

P.S. 50 runs done for all comparisons.

**Comment by joergoster** [comment]
@lucasart Let me clarify: if we would do let's say 4 or 5 profile builds for each branch, and then choose the fastest, I would not object.

**Comment by lucasart** [comment]
You are changing the subject. And regarding profile build, you are just seeing what you **want** to see. What if you run your test with pb1 twice ? Of course you get different results...

And there is certainly something wrong with your results. Your "variance of the mean" -- which I assume is the "standard deviation of the mean" -- is way too high to be credible. With only 20 runs, I get `0.26%` on my machine. Are you forgetting to divide by sqrt(N) somewhere ? How many runs do you do to compute these stats ?

File: src/search.cpp [hunk 1]
```diff
@@ -596,19 +596,20 @@ namespace {
     int moveCount, quietCount;
 
     // Step 1. Initialize node
-    Thread* thisThread = pos.this_thread();
+    prefetch(ThisThread->materialTable[pos.material_key()]);
+    prefetch(ThisThread->pawnsTable[pos.material_key()]);
```
Comment (mcostalba) [inline src/search.cpp]: Theseprefetches have nothing to do wtith this patch. Why did you add them here?

File: src/thread.cpp [hunk 2]
```diff
@@ -27,7 +27,8 @@
 
 using namespace Search;
 
-ThreadPool Threads; // Global object
+ThreadPool Threads;
+thread_local Thread *ThisThread;
```
Comment (mcostalba) [inline src/thread.cpp]: thread_local is a nice solution, but the last time I tested it, I found very slow:

http://stackoverflow.com/questions/506093/why-is-thread-local-storage-so-slow

BTW that's one of the reasons why thread is in position.

Comment (mcostalba) [inline src/thread.cpp]: See here:
https://software.intel.com/en-us/blogs/2011/05/02/the-hidden-performance-cost-of-accessing-thread-local-variables

"In order to understand where the cost of accessing thread local variable comes from, I needed to understand how a compiler implements it. The compiler will assign a unique global ID to each thread-local variable (this ID is the same for each thread), and maintains a Thread Local Storage (TLS) lookup table for each thread. The global ID then used to find the address of any thread local variable. So the cost of using a thread-local variable includes the cost of a function call and a lookup in the indexed table. "

File: src/search.cpp (hunk 1 above)
Comment (lucasart) [inline src/search.cpp]: If you look  carefully, you'll  see  that I had to move  the perfecth from  do_move

**Comment by lucasart** [comment]
@mcostalba: did you  run  some  speed  measures  on  this  patch ? there's  no  slowdown  on  my  machine, at  least.

**Comment by mcostalba** [comment]
I ran almost one year ago when I was exploring the possibility to use
thread_local. I never tested it again since then.

File: src/search.cpp (hunk 1 above)
Comment (lucasart) [inline src/search.cpp]: i'll move  them  back  in  do_move and  run  some  speed  test.

**Comment by lucasart** [comment]
To investigate Joerg's claim, I redid the experiment, redoing the profile build on both master and test branch:
* `make profile-build ARCH=x86-64-modern && make strip`, on both branches.
* 20 bench for both branches, running both versions concurrently, to reduce the noise on the NPS diff vector.
No significant difference observed. Still a tiny (but statistically significant) speed-up on my machine:
```
stat        test     master   diff
mean   2,590,165  2,585,254  6,829
stdev     17,994     16,162  9,379

speedup       0.26%
P(speedup>0)  99.9%
```
Here are the 20 run results (in NPS), with the corresponding NPS diff vector (so you can see the noise is extremely small):
```
test	master	diff
2,541,205	2,525,078	16,127
2,590,044	2,579,649	10,395
2,604,577	2,592,455	12,122
2,545,851	2,569,337	-23,486
2,590,848	2,589,242	1,606
2,570,127	2,583,637	-13,510
2,590,044	2,583,637	6,407
2,594,065	2,589,242	4,823
2,603,765	2,604,577	-812
2,591,651	2,582,041	9,610
2,603,765	2,589,242	14,523
2,594,065	2,594,065	0
2,599,715	2,586,837	12,878
2,598,098	2,590,848	7,250
2,597,290	2,588,440	8,850
2,582,839	2,579,649	3,190
2,604,577	2,599,715	4,862
2,601,334	2,596,483	4,851
2,598,098	2,586,837	11,261
2,601,334	2,594,065	7,269
```
I'm not making this up! Here's my shell script:
```
#!/bin/bash
sleep 0.5s
rm $1.csv
for (( i = 1; i <= 20; i++ ))
do
    echo "$i/20"
    x=$($1 bench 2>&1|grep "Nodes\/second"|cut -d\: -f2)
    echo $x >> $1.csv
done
```
You need to run both versions concurrently, like so:
```
$ ./bench ./Stockfish/master & ./bench ./Stockfish/test
```
This is important to reduce the noise of the diff vector (see above). Concurrency ensures that both versions share (approx.) the same level of external noise in each run, thereby reducing the variance of the diff: `V(diff) < V(test) + V(master)`. Instead, if you run both one after the other, you have independance, hence decorrelation, and `V(diff) = V(test) + V(master)`.

File: src/search.cpp (hunk 1 above)
Comment (lucasart) [inline src/search.cpp]: @mcostalba: I tried to move back prefetch in do_move(), and got no speed-up. even a minor slowdown (-0.3%). The reason it does not belong in do_move() is because threads don't belong in position. So it makes sense to move them out of do_move() and into search instead.

**Comment by mstembera** [comment]
I like this patch unfortunately it's a BIG around 6% to 7% slowdown for me under Windows64 for both non profile and profile builds.  gcc 5.2.0 measured using Fishbench.

Non profile:
make build ARCH=x86-64-modern COMP=mingw
```
Results for 20 tests for each version:

            Base      Test      Diff      
    Mean    1086811   1013175   73636     
    StDev   65053     60245     7454      

p-value: 0
speedup: -0.068
```

Profile:
make profile-build ARCH=x86-64-modern COMP=mingw
```
Results for 20 tests for each version:

            Base      Test      Diff      
    Mean    1196026   1118375   77651     
    StDev   34067     27771     7232      

p-value: 0
speedup: -0.065
```

**Comment by lucasart** [comment]
@mstembera: Interesting. I wonder if the problem is with Windows, or MinGW. There is no reason why this should be any slower than master, other than the fact that MinGW (or Windows) sucks with 2 straws...

**Comment by lucasart** [comment]
I will check this in fishtest. It needs to be validated in SMP as well, not just for performance in single threaded case. It should be a non functional change, even in SMP, but with SMP one is never too careful. Fishtest will also show clearly if the problem only affects Windows machines.

**Comment by mstembera** [comment]
@lucasart 
I can't figure out what the reason is either(except what Marco already mentioned regarding thread_local).  I agree your plan to see what happens on fishtest will be informative.

**Comment by lucasart** [comment]
@mstembera: I just pushed another patch. Can you rerun your Windows measures ?

Patch simply caches `thisThread` at the beginning of the search, to avoid TLS lookups that are so expensive on Windows. I hope it helps, but can't measure it, as I don't have Windows.

On my Linux machine, the patch is a significant slowdown, but still not far from master: `-0.34% +/-0.04%`. So it could be a reasonable compromise.

**Comment by mcostalba** [comment]
I have pushed a patch above your series:

https://github.com/mcostalba/Stockfish/commit/thisThread

Here I revert to using a std::map to lookup the thread pointer. It i based on std::this_thread::id that returns the id of the thread that is running the code at the caller site.

**Comment by mcostalba** [comment]
My previous patch is just an example. I am not advocating it for merging. Indeed it is anyhow slower than current code. Actually if we cannot use thread_local then the fastest solution is to store this_thread pointer in an object that is:

1) Accessible everywhere this_thread is needed
2) There is one and only one for each thread

This yields to storing this_thread in Position, and that's why it is there!

**Comment by lucasart** [comment]
So TLS can't be used, because Windows' implementation of TLS sucks dead donkeys through a straw ?

That's really frustrating. But there's not much we can do about it. Just another fact of life...

Closing the pull request.

**Comment by mstembera** [comment]
I retested your new version anyway just to be sure but it made little difference.
```
Results for 20 tests for each version:

            Base      Test      Diff      
    Mean    1079398   1016159   63239     
    StDev   14849     13386     2089      

p-value: 0
speedup: -0.059
```
```
Results for 20 tests for each version:

            Base      Test      Diff      
    Mean    1220309   1136543   83766     
    StDev   53496     50170     7316      

p-value: 0
speedup: -0.069
```

**Comment by lucasart** [comment]
@mstembera: thanks for checking.
